device/cuda/CMakeLists.txt

```diff
@@ -111,115 +111,43 @@ traccc_add_library( traccc_cuda cuda TYPE SHARED
-                "${TEMPLATE_SOURCE}"
-                -o "${GENERATED_SOURCE}"
-                --detector "${DETECTOR_NAME}"
-                --bfield "${BFIELD_NAME}"
-                --model cuda
-            DEPENDS "${KERNEL_SPECIALIZATION_PY}" "${TEMPLATE_SOURCE}"
-            COMMENT "Generating kernel specialization of `fit_forward` for ${DETECTOR_NAME} and ${BFIELD_NAME}"
-        )
-        target_sources(traccc_cuda PRIVATE ${GENERATED_SOURCE})
-    endforeach()
-endforeach()
-
-# Generate specializations for fit_backward
+# Create a well formed postfix for the specialized filenames.
+function(traccc_make_cuda_fname_postfix FNAME)
+  set("${FNAME}" "")
+  if(NOT "${DETECTOR_NAME}" STREQUAL "")
+    set("${FNAME}" "${${FNAME}}.${DETECTOR_NAME}")
+  endif()
+  if(NOT "${BFIELD_NAME}" STREQUAL "")
+    set("${FNAME}" "${${FNAME}}.${BFIELD_NAME}")
+  endif()
+  string(REPLACE "<scalar>" ""  "${FNAME}" "${${FNAME}}")
+  set("${FNAME}" "${${FNAME}}" PARENT_SCOPE)
+endfunction()
+
+# Helper macro for adding a kernel specialization to the build of traccc::cuda.
+function(traccc_add_cuda_specialization TEMPLATE_FILE)
+  traccc_make_cuda_fname_postfix(FNAME_POSTFIX)
+  configure_file("${TEMPLATE_FILE}" "${TEMPLATE_FILE}${FNAME_POSTFIX}.cu")
+  target_sources(traccc_cuda PRIVATE
+    "${CMAKE_CURRENT_BINARY_DIR}/${TEMPLATE_FILE}${FNAME_POSTFIX}.cu")
+endfunction()
+
+# Generate specializations for the kernels.
 foreach(DETECTOR_NAME ${TRACCC_SUPPORTED_DETECTORS})
-    foreach(BFIELD_NAME ${TRACCC_CUDA_SUPPORTED_BFIELDS})
-        set(GENERATED_SOURCE "${CMAKE_CURRENT_BINARY_DIR}/src/fitting/kernels/specializations/fit_backward_${DETECTOR_NAME}_${BFIELD_NAME}.cu")
-        set(TEMPLATE_SOURCE "${CMAKE_CURRENT_SOURCE_DIR}/src/fitting/kernels/specializations/fit_backward.cu.template")
-        add_custom_command(
-            OUTPUT "${GENERATED_SOURCE}"
-            COMMAND ${CMAKE_COMMAND} -E make_directory "${CMAKE_CURRENT_BINARY_DIR}/src/fitting/kernels/specializations/"
-            COMMAND
-                Python::Interpreter
-                "${KERNEL_SPECIALIZATION_PY}"
-                "${TEMPLATE_SOURCE}"
-                -o "${GENERATED_SOURCE}"
-                --detector "${DETECTOR_NAME}"
-                --bfield "${BFIELD_NAME}"
-                --model cuda
-            DEPENDS "${KERNEL_SPECIALIZATION_PY}" "${TEMPLATE_SOURCE}"
-            COMMENT "Generating kernel specialization of `fit_backward` for ${DETECTOR_NAME} and ${BFIELD_NAME}"
-        )
-        target_sources(traccc_cuda PRIVATE ${GENERATED_SOURCE})
-    endforeach()
+  traccc_add_cuda_specialization(
+    "src/finding/kernels/specializations/find_tracks.cu.in")
+  traccc_add_cuda_specialization(
+    "src/finding/kernels/specializations/apply_interaction.cu.in")
+  foreach(BFIELD_NAME "const_bfield_backend_t<scalar>"
+                      "inhom_global_bfield_backend_t<scalar>"
+                      "inhom_texture_bfield_backend_t")
```

Comment:
Bring back the list variable here, and as mentioned before the `<scalar>` template argument needs to go.

Reply:
I'm inputting C++ types here. 🤔 I don't understand your objection. `scalar` is a valid replacement in your template files.